training_system/spec/controller/trainer/courses_controller_spec.rb

```diff
@@ -0,0 +1,211 @@
+require "rails_helper"
+include SessionsHelper
+
+RSpec.describe Trainer::CoursesController, type: :controller do
+  let!(:user) {FactoryBot.create :user}
+  let!(:course) {FactoryBot.create :course}
+
+  describe "GET #index" do
+    context "when user is loggin" do
+      before do
+        log_in user
+        get :index
+      end
+
+      it "assign @courses" do
+        expect(assigns(:courses)).to eq([course])
+      end
+    end
+
+    context "fail when user login not yet" do
+      before do
+        get :index
+      end
+
+      it "show flash danger" do
+        expect(flash[:danger]).to eq I18n.t("controllers.course_controller.let_login")
+      end
+
+      it "redirect to login_path" do
+        expect(response).to redirect_to login_path
+      end
+    end
+  end
+
+  describe "GET new" do
+    context "when user is loggin render new" do
+      before do
+        log_in user
+        get :new
+      end
+
+      it "should render #new template" do
+        expect(response).to render_template(:new)
+      end
+    end
+  end
+
+  describe "GET #show" do
+    context "when user is loggin" do
+      before do
+        log_in user
+        get :show, params: { id: course.id }
+      end
+
+      it "should show course" do
+        expect(assigns(:course)).to eq(course)
+      end
+
+      it "renders the #show view" do
+        expect(response).to render_template :show
+      end
+    end
+  end
+
+  describe "GET #edit" do
+    context "when user is loggin" do
+      before do
+        log_in user
+        get :edit, params: { id: course.id }
+      end
+
+      it "should show course" do
+        expect(assigns(:course)).to eq(course)
+      end
+
+      it "renders the #edit view" do
+        expect(response).to render_template :edit
+      end
+    end
+  end
+
+  describe "POST course#create" do
+    let!(:user1) {FactoryBot.create :user}
+    let!(:user2) {FactoryBot.create :user}
+    let!(:subject1) {FactoryBot.create :subject}
+    let!(:subject2) {FactoryBot.create :subject}
+    context "success when valid attributes" do
+      before do
+        log_in user
+        get :create, params: {
+          course: {
+            name: "ruby",
+            description: "test",
+            start_date: "2021-09-03",
+            end_date: "2021-09-30",
+            status: :init,
+            user_ids: [user1.id, user2.id],
+            subject_ids: [subject1.id, subject2.id]
+          }
```

Comment:
chỗ ni nên khai báo ra 1 let rồi truyền vào fen

Reply:
thiệp nói đúng nè em, cho nó gọn code, dùng `let`

Reply:
ngoài ra, chị chả thấy em check xem là đúng là DB đã tăng lên 1 record chưa j trơn 
`it "save success with valid data" do`
  `        expect do`
   `         post :create, params: {course: params_valid, user_id: user.id}` (chỗ parram này em tự set cho đúng à dc, chị ví dụ thôi 
 `         end.to change(Course, :count).by 1`
`        end`